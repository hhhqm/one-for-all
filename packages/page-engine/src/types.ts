import { Serialized, BaseNode, NodeType, APIStatesSpec, SharedStatesSpec } from '@ofa/render-engine';

export type ReactComp = React.ComponentType | React.JSXElementConstructor<any>;

export interface PageNode extends BaseNode<Serialized> {
  id: string;
  pid?: string; // only used on page-engine
  type: NodeType.ReactComponentNode | NodeType.LoopContainerNode | NodeType.HTMLNode;
  label: string;
  name?: string; // for html node
  // `packageName, packageVersion, exportName` only for react comp node
  packageName?: 'ofa-ui' | string;
  packageVersion?: 'latest' | string;
  exportName: 'page' | 'elemName' | string; // registry elem type
  children?: Array<PageNode>
}

export interface PageSchema {
  node: PageNode;
  apiStateSpec: APIStatesSpec;
  sharedStatesSpec: SharedStatesSpec;
}

export interface iconStyle {
  width: string,
  height: string
}

export type DragPos = 'up' | 'down' | 'left' | 'right' | 'inner';

// registry types
export type Category = 'basic' | 'layout' | 'form' | 'advanced' | string;

export interface SourceElement<T> {
  name: string;
  component: ReactComp;
  defaultConfig: Record<string, any>; // 表单默认配置
  configForm: ReactComp; // 属性配置组件
  icon: string;
  label: string;
  category: Category;
  iconStyle?: iconStyle;
  iconSize?: number;
  toProps?: (formData: any) => T; // 将configForm的配置项转换到 component的 props
  order?: number; // 排序权重
  hidden?: boolean; // 在source panel 隐藏
  acceptChild?: boolean; // 是否接受子节点
  exportActions?: string[]; // 对外暴露的方法名
  defaultStyle?: Record<string, any>; // 默认样式
}
