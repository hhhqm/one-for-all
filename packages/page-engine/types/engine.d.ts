type ReactComp = React.ComponentType | React.JSXElementConstructor<any>;
type PrimitiveType = string | number | boolean | undefined | null

declare module '*.m.scss' {
  const classes: { [key: string]: string };
  export default classes;
}

declare namespace PageEngine {
  interface Node {
    comp: string;
    id?: string;
    pid?: string;
    label?: string; // 在page tree展示的名称
    props?: Record<string, any> | null;
    _shared?: Record<string, any>; // 页面的普通变量数据源, 只存于page节点
    _api?: Record<string, any>; // 页面的api变量数据源, 只存于page节点
    _stateRef?: Record<string, any>; // 节点引用的数据源的key path, 包括_shared, _api 数据, 存于page 以下的任何节点
    _style?: React.CSSProperties; // setting panel的style配置
    _events?: Record<string, any>; // setting panel的 events配置
    _renderer?: Record<string, any>; // setting panel的 动态渲染配置
    children?: Array<Node | PrimitiveType>;
  }

  type DragPos = 'up' | 'down' | 'left' | 'right' | 'inner';
}
