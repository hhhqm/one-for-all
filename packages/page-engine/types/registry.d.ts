declare namespace Registry {
  type Category = 'basic' | 'layout' | 'form' | 'advanced' | string;

  interface SourceElement<T> {
    name: string;
    component: ReactComp;
    defaultConfig: Record<string, any>; // 表单默认配置
    configForm: ReactComp; // 属性配置组件
    icon: string;
    label: string;
    category: Category;
    toProps?: (formData: any) => T; // 将configForm的配置项转换到 component的 props
    order?: number; // 排序权重
    hidden?: boolean; // 在source panel 隐藏
    acceptChild?: boolean; // 是否接受子节点
    expose?: Record<string, any>; // 对外暴露的属性/方法
    defaultStyle?: Record<string, any>; // 样式默认配置
  }
}
