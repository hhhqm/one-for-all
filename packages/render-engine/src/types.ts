import React from 'react';
import type { BehaviorSubject } from 'rxjs';
import type { FetchParams, APISpecAdapter } from '@ofa/api-spec-adapter';

export type Serialized = 'Serialized';
export type Instantiated = 'Instantiated';

// APIState define the type of API results from view perspective.
// This type is inspired by [react-query](https://react-query.tanstack.com/).
export type APIState = {
  loading: boolean;
  result?: unknown;
  error?: Error;
};

export const enum NodePropType {
  ConstantProperty = 'constant_property',
  APIResultProperty = 'api_result_property',
  APILoadingProperty = 'api_loading_property',
  // todo api error property
  SharedStateProperty = 'shared_state_property',
  NodeStateProperty = 'node_state_property',

  /**
   * @deprecated This type has been deprecated, please use FunctionalProperty instead
   */
  APIInvokeProperty = 'api_invoke_property',
  SharedStateMutationProperty = 'shared_state_mutation_property',
  FunctionalProperty = 'functional_property',
  RenderProperty = 'render_property',
}

export type NodeProperty<T extends Serialized | Instantiated> =
  ConstantProperty |
  APIResultProperty<T> |
  APILoadingProperty |
  SharedStateProperty<T> |
  NodeStateProperty<T> |
  FunctionalProperty<T> |
  SharedStateMutationProperty<T> |
  APIInvokeProperty<T> |
  RenderProperty<T>;
  // Array<APIInvokeProperty<T>>;

export type NodeProperties<T extends Serialized | Instantiated> = Record<string, NodeProperty<T>>;

type BaseNodeProperty = {
  type: NodePropType;
}

export type ConstantProperty = BaseNodeProperty & {
  type: NodePropType.ConstantProperty;
  value: unknown;
}

export type APIResultProperty<T> = BaseNodeProperty & {
  type: NodePropType.APIResultProperty;
  stateID: string;
  // Convertor is used to transform the API result before passing it to node,
  // convertor will never be called if API request failed or the result is nullish.
  // the signature of this function is: (state: unknown) => unknown,
  // there is only one argument called `state`, which is the `result` in APIState
  convertor?: StateConvertor<T>;
  // It is a best practice to always define a fallback for API results,
  // no matter before the API response returned or after an unexpected error has occurred.
  //
  // fallback should be a NOT nullish value, and will be passed to node in the following situations:
  // - the initial state
  // - API request failed or some business error returned
  // - convertor throw an error when calling it with API result
  // - convertor return `null` or `undefined`
  //
  // The value of fallback will not always be the same.
  // It will be assign a new value when convertor returned a not-nullish value.
  // If there is no convertor defined, fallback will be assigned to the latest not-nullish api result.
  //
  // Be Attention. fallback is NOT the fallback of API result, it is the fallback of a property passed to node.
  fallback: unknown;
}

export type StateConvertor<T> = T extends Serialized ? SerializedStateConvertor : StateConvertorFunc;
export type SerializedStateConvertor = StateConvertExpression | StateConvertorFuncSpec;
export type StateConvertorFunc = (v: any) => any;

export type StateConvertExpression = {
  type: 'state_convert_expression';
  // same as JavaScript expression, with a predefined variable called `state`
  // state.foo
  // state.offset / (state.limit + state.offset)
  // state.list.map((item) => item.name))
  // state.list.map((item) => `名称：${item.name}`))
  // state.foo?.bar?.baz || 'someValue'
  expression: string;
}

export type StateConvertorFuncSpec = BaseFunctionSpec & {
  type: 'state_convertor_func_spec';
  args: 'state';
};

// toProps function should return Record<string, unknown>;
export type ToPropsFuncSpec = BaseFunctionSpec & {
  type: 'to_props_function_spec',
  args: 'state'
};
export type ToProps<T> = T extends Serialized ?
  ToPropsFuncSpec :
  (state: unknown) => Record<string, unknown>;

export type APILoadingProperty = BaseNodeProperty & {
  type: NodePropType.APILoadingProperty;
  stateID: string;
}

export type SharedStateProperty<T> = BaseNodeProperty & {
  type: NodePropType.SharedStateProperty;
  stateID: string;
  fallback: unknown;
  convertor?: StateConvertor<T>;
}

export type NodeStateProperty<T> = BaseNodeProperty & {
  type: NodePropType.NodeStateProperty;
  nodeKey: string;
  fallback: unknown;
  convertor?: StateConvertor<T>;
}

export type FunctionalProperty<T> = BaseNodeProperty & {
  type: NodePropType.FunctionalProperty;
  func: T extends Serialized ? BaseFunctionSpec : VersatileFunc;
}

// todo refactor this type property spec
export type SharedStateMutationProperty<T> = {
  type: NodePropType.SharedStateMutationProperty;
  stateID: string;
  convertor?: T extends Serialized ? BaseFunctionSpec : VersatileFunc;
}

/**
 * @deprecated This type has been deprecated, please use FunctionalProperty instead
 */
export type APIInvokeProperty<T> = {
  type: NodePropType.APIInvokeProperty;
  stateID: string;
  // the required return type is too complex
  paramsBuilder?: ParamsBuilder<T>;
  callback?: T extends Serialized ? APIFetchCallbackSpec : APIFetchCallback;
}

// It is common to passing a component to another component,
// or defined a function which may take some argument and return a component.
// RenderProperty is the spec for this kind of property.
//
// <ParentComponent
//   render={<SomeComponent />}
// />
// <ParentComponent
//   render={(someData): JSX.Element => (<SomeComponent data={someData} />)}
// />
// <ParentComponent
//   render={(someData, someIgnoredValue): JSX.Element => (<SomeComponent data={someData} otherProp={otherProp} />)}
// />
export type RenderProperty<T extends Serialized | Instantiated> = BaseNodeProperty & {
  type: NodePropType.RenderProperty;
  adapter: RenderPropertyAdapter<T>;
  node: SchemaNode<T>;
}

export type RenderPropertyAdapterFuncSpec = BaseFunctionSpec & {
  type: 'render_property_function_spec',
};

export type RenderPropertyAdapter<T> = T extends Serialized ?
  RenderPropertyAdapterFuncSpec :
  (...args: unknown[]) => Record<string, unknown>;

export type ParamsBuilder<T> = T extends Serialized ?
  ParamsBuilderFuncSpec : (...args: unknown[]) => FetchParams;

export type ParamsBuilderFuncSpec = BaseFunctionSpec & {
  type: 'param_builder_func_spec';
}

export type LifecycleHookFuncSpec = BaseFunctionSpec & {
  type: 'lifecycle_hook_func_spec';
  args: '';
}

export type BaseFunctionSpec = {
  type: string;
  args: string;
  body: string;
}

export type FetchOption = {
  params?: FetchParams;
  // Callback is the hook for performing side effect after an API request.
  //
  // Why use callback style hook instead of Promise style?
  //
  // Every node should be equal, but some nodes are more equal than others.
  // In Pro-Code scenario, we can carefully design a Model to handle view actions and perform API request,
  // we also can carefully design some FLAG state to prevent extra API request or an complex cancellation mechanism.
  // These solutions may not be difficult for experienced programmers,
  // but our goal of building low-code platform is to enable junior programmers, even non-programmers,
  // to develop complex applications quickly and easily.
  // To solve this problem, we use `rxjs/ajax` to implement automatic cancellation of requests.
  // Doing so led us to a small problem, we can not use Promise style hook.
  // As said at the beginning, every node should be equal,
  // so every node could perform a side effect after some API request,
  // and if we use a Promise style hook,
  // there must will be some hooks will never be called and stay in pending status forever,
  // just like the following code:
  //
  //   new Promise((resolve, reject) => {
  //     setTimeout(() => {
  //       // resolve will never be called
  //       resolve();
  //     }, forever);
  //   });
  //
  // We, the developers, design a action which has a side effect,
  // but user make the finial decision when to trigger that action.
  // Side effect could be perform as many times as wanted,
  // but only under user's expectation.
  // When user trigger the *same action* many times,
  // only the last one should perform side effect.
  //
  // ps. Same actions means they have the same API stateID.
  callback?: APIFetchCallback;
}

export interface StatesHubAPI {
  hasState$: (stateID: string) => boolean;
  findState$: (stateID: string) => APIState$WithActions | undefined;
  getState$: (stateID: string) => BehaviorSubject<APIState>;
  fetch: (stateID: string, fetchOption: FetchOption) => void;
  refresh: (stateID: string) => void;
}

export type APIState$WithActions = {
  state$: BehaviorSubject<APIState>;
  fetch: (fetchOption: FetchOption) => void;
  refresh: () => void;
};

export type APIFetchCallbackSpec = BaseFunctionSpec & {
  type: 'api_fetch_callback';
  args: '{ result, error }',
}

export type APIFetchCallback = (state: Omit<APIState, 'loading'>) => void;
export type APIFetch = (params: FetchParams, callback?: APIFetchCallback) => void;

export interface StatesHubShared {
  hasState$: (stateID: string) => boolean;
  findState$: (stateID: string) => BehaviorSubject<unknown> | undefined;
  getState$: (stateID: string) => BehaviorSubject<unknown>;
  getNodeState$: (nodeKey: string) => BehaviorSubject<unknown>;
  exposeNodeState: (nodeKey: React.Key, state: unknown) => void;
  mutateState: (stateID: string, state: unknown) => void;
}

export type APIStateWithFetch = APIState & {
  fetch: APIFetch;
  refresh: () => void;
}

export type CTX = {
  statesHubAPI: StatesHubAPI;
  statesHubShared: StatesHubShared;
  apiStates: Readonly<Record<string, APIStateWithFetch>>;
  states: Record<string, unknown>;
  repository?: Repository;
  refLoader?: RefLoader;
}

export type RenderEngineCTX = Pick<CTX, 'states' | 'apiStates'>;

export type VersatileFunc<T = unknown> = (...args: unknown[]) => T;

export type LifecycleHooks<T extends Serialized | Instantiated> = Partial<{
  didMount: T extends Serialized ? LifecycleHookFuncSpec : VersatileFunc;
  willUnmount: T extends Serialized ? LifecycleHookFuncSpec : VersatileFunc;
}>;

export const enum NodeType {
  HTMLNode = 'html-element',
  ReactComponentNode = 'react-component',
  LoopContainerNode = 'loop-container',
  ComposedNode = 'composed-node',
  RefNode = 'ref-node',
}

export interface BaseNode<T extends Serialized | Instantiated> {
  id: React.Key;
  type: NodeType;
  props?: NodeProperties<T>;
  lifecycleHooks?: LifecycleHooks<T>;
}

export interface HTMLNode<T extends Serialized | Instantiated> extends BaseNode<T> {
  type: NodeType.HTMLNode;
  name: string;
  children?: Array<SchemaNode<T>>;
}

export interface ReactComponentNode<T extends Serialized | Instantiated> extends BaseNode<T> {
  type: NodeType.ReactComponentNode;
  packageName: string;
  packageVersion: string;
  exportName: 'default' | string;
  supportStateExposure?: boolean;
  // not recommend, should avoid
  children?: Array<SchemaNode<T>>;
}

export type PlainState<T extends Serialized | Instantiated> =
  APIResultProperty<T> |
  SharedStateProperty<T> |
  NodeStateProperty<T> |
  ConstantProperty;

export interface LoopContainerNode<T extends Serialized | Instantiated> extends BaseNode<T> {
  type: NodeType.LoopContainerNode;
  // props: LoopContainerNodeProps<T>;
  iterableState: PlainState<T>;
  loopKey: string;
  node: SchemaNode<T>;
  toProps: ToProps<T>;
}

export type ComposedNodeChild<T extends Serialized | Instantiated> = SchemaNode<T> & {
  toProps?: ToProps<T>;
}

export interface ComposedNode<T extends Serialized | Instantiated> extends BaseNode<T> {
  type: NodeType.ComposedNode;
  outLayer?: Omit<HTMLNode<T>, 'children'>;
  composedState: PlainState<T>;
  children: Array<ComposedNodeChild<T>>;
}

export interface RefNode<T extends Serialized | Instantiated> extends BaseNode<T> {
  type: NodeType.RefNode;
  schemaID: string;
  fallback?: SchemaNode<T>;
  // RefNode will inherit parent context,
  // which means use states if not found in current context.
  // set `orphan` to `true` to disable inheritance
  orphan?: boolean;
}

export type SchemaNode<T extends Serialized | Instantiated> =
  HTMLNode<T> |
  ReactComponentNode<T> |
  LoopContainerNode<T> |
  ComposedNode<T> |
  RefNode<T>;

// map of stateID and apiID
// todo should also store builder info
export type APIStatesSpec = Record<string, { apiID: string; [key: string]: unknown; }>;

export type SharedStatesSpec = Record<string, { initial: unknown; }>;

export type Schema = {
  node: SchemaNode<Serialized>;
  apiStateSpec?: APIStatesSpec;
  sharedStatesSpec?: SharedStatesSpec;
}

export type InstantiatedNode = SchemaNode<Instantiated>;

// todo fixme
export type DynamicComponent = React.FC<any> | React.ComponentClass<unknown>;

type PackageNameVersion = string;
export type Repository = Record<PackageNameVersion, Record<string, DynamicComponent>>;

export type RefLoader = (schemaID: string) => Promise<InitProps>;

export type InitProps = {
  schema: Schema;
  apiSpecAdapter: APISpecAdapter;
  repository?: Repository;
  refLoader?: RefLoader;
}
