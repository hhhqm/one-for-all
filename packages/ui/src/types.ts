export interface GridProps {
  colRatio: string; // 列比例
  colGap: string; // 列间距
  className?: string;
  style?: React.CSSProperties;
  children?: React.ReactNode;
}

export interface InputProps {
  style?: React.CSSProperties;
  children?: React.ReactNode;
  placeholder?: string;
  type?: string;
}
