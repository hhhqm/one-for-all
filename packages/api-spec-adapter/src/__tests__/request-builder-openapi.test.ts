import spec from './fixtures/petstore-spec';
import SwaggerAdapter from '../adapter-swagger';
import { RequestParams } from '../types';

const actions: Array<{ path: string; method: string; requestParams?: RequestParams; }> = [
  {
    path: '/pet',
    method: 'post',
    requestParams: { body: { foo: 'bar' } },
  },
  {
    path: '/pet',
    method: 'put',
    requestParams: { body: { foo: 'bar' } },
  },
  {
    path: '/pet/findByStatus',
    method: 'get',
    requestParams: { params: { status: ['available', 'pending', 'sold'] } },
  },
  {
    path: '/pet/findByTags',
    method: 'get',
    requestParams: { params: { tags: ['tag1', 'tag2', 'tag3'] } },
  },
  {
    path: '/pet/{petId}',
    method: 'get',
    requestParams: { params: { petId: 'some_pet_id' } },
  },
  {
    path: '/pet/{petId}',
    method: 'post',
    requestParams: { params: { petId: 'some_pet_id' }, body: { name: 'foo', status: 'alive' } },
  },
  {
    path: '/pet/{petId}',
    method: 'delete',
    requestParams: { params: { petId: 'some_pet_id' } },
  },
  {
    path: '/pet/{petId}/uploadImage',
    method: 'post',
    requestParams: { params: { petId: 'some_pet_id' } },
  },
  {
    path: '/store/inventory',
    method: 'get',
  },
  {
    path: '/store/order',
    method: 'post',
    requestParams: { body: { foo: 'bar' } },
  },
  {
    path: '/store/order/{orderId}',
    method: 'get',
  },
  {
    path: '/store/order/{orderId}',
    method: 'delete',
  },
  {
    path: '/user',
    method: 'post',
  },
  {
    path: '/user/createWithArray',
    method: 'post',
  },
  {
    path: '/user/createWithList',
    method: 'post',
  },
  {
    path: '/user/login',
    method: 'get',
  },
  {
    path: '/user/logout',
    method: 'get',
  },
  {
    path: '/user/{username}',
    method: 'get',
  },
];

test('expect_specInterpreter_to_match_snapshots', () => {
  const builder = new SwaggerAdapter(spec);
  actions.forEach(({ path, method, requestParams }) => {
    expect(builder.build(`${method}:${path}`, requestParams)).toMatchSnapshot();
  });
});
