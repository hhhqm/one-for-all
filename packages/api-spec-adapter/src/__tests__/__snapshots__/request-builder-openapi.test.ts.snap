// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect_specInterpreter_to_match_snapshots 1`] = `
Object {
  "body": Object {
    "foo": "bar",
  },
  "headers": Object {},
  "method": "post",
  "queryParams": Object {},
  "url": "/pet",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 2`] = `
Object {
  "body": Object {
    "foo": "bar",
  },
  "headers": Object {},
  "method": "put",
  "queryParams": Object {},
  "url": "/pet",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 3`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "get",
  "queryParams": Object {
    "status": Array [
      "available",
      "pending",
      "sold",
    ],
  },
  "url": "/pet/findByStatus",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 4`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "get",
  "queryParams": Object {
    "tags": Array [
      "tag1",
      "tag2",
      "tag3",
    ],
  },
  "url": "/pet/findByTags",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 5`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "get",
  "queryParams": Object {},
  "url": "/pet/some_pet_id",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 6`] = `
Object {
  "body": Object {
    "name": "foo",
    "status": "alive",
  },
  "headers": Object {},
  "method": "post",
  "queryParams": Object {},
  "url": "/pet/some_pet_id",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 7`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "delete",
  "queryParams": Object {},
  "url": "/pet/some_pet_id",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 8`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "post",
  "queryParams": Object {},
  "url": "/pet/some_pet_id/uploadImage",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 9`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "get",
  "queryParams": Object {},
  "url": "/store/inventory",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 10`] = `
Object {
  "body": Object {
    "foo": "bar",
  },
  "headers": Object {},
  "method": "post",
  "queryParams": Object {},
  "url": "/store/order",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 11`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "get",
  "queryParams": Object {},
  "url": "/store/order/{orderId}",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 12`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "delete",
  "queryParams": Object {},
  "url": "/store/order/{orderId}",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 13`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "post",
  "queryParams": Object {},
  "url": "/user",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 14`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "post",
  "queryParams": Object {},
  "url": "/user/createWithArray",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 15`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "post",
  "queryParams": Object {},
  "url": "/user/createWithList",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 16`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "get",
  "queryParams": Object {},
  "url": "/user/login",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 17`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "get",
  "queryParams": Object {},
  "url": "/user/logout",
}
`;

exports[`expect_specInterpreter_to_match_snapshots 18`] = `
Object {
  "body": undefined,
  "headers": Object {},
  "method": "get",
  "queryParams": Object {},
  "url": "/user/{username}",
}
`;
